File: netlify/functions/movers.py
import json, math

def safe(val):
    if val is None: return None
    if isinstance(val, float) and (math.isnan(val) or math.isinf(val)): return None
    return val

def handler(event, context):
    headers = {"Access-Control-Allow-Origin": "*", "Content-Type": "application/json"}
    try:
        import yfinance as yf
        symbols = ["NVDA","TSLA","AMD","COIN","PLTR","MSTR","SMCI","IONQ","RKLB","SHOP"]
        result = []
        for sym in symbols:
            try:
                info = yf.Ticker(sym).info or {}
                price = safe(info.get("regularMarketPrice") or info.get("currentPrice"))
                prev = safe(info.get("regularMarketPreviousClose") or info.get("previousClose"))
                chg = ((price-prev)/prev*100) if price and prev else None
                result.append({"ticker":sym,"name":info.get("shortName") or sym,"price":price,"change":safe(chg)})
            except: pass
        result.sort(key=lambda x: x["change"] or 0, reverse=True)
        gainers = [r for r in result if (r["change"] or 0) > 0][:5]
        losers = sorted([r for r in result if (r["change"] or 0) < 0], key=lambda x: x["change"] or 0)[:5]
        return {"statusCode": 200, "headers": headers, "body": json.dumps({"gainers":gainers,"losers":losers})}
    except Exception as e:
        return {"statusCode": 500, "headers": headers, "body": json.dumps({"gainers":[],"losers":[],"error":str(e)})}

File: netlify/functions/movers.js
const https = require("https");

function get(url) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, { timeout: 12000 }, (res) => {
      let data = "";
      res.on("data", (c) => (data += c));
      res.on("end", () => {
        try {
          resolve(JSON.parse(data));
        } catch {
          reject(new Error("JSON parse error"));
        }
      });
    });
    req.on("error", reject);
    req.on("timeout", () => {
      req.destroy();
      reject(new Error("timeout"));
    });
  });
}

function safe(v) {
  if (v === null || v === undefined) return null;
  if (typeof v === "number" && (!isFinite(v) || isNaN(v))) return null;
  return v;
}

exports.handler = async (event) => {
  const headers = {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  };
  const KEY = process.env.FINNHUB_API_KEY || process.env.finnhub_api_key;
  if (!KEY)
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({
        gainers: [],
        losers: [],
        error: "FINNHUB_API_KEY not set",
      }),
    };

  try {
    const symbols = [
      "NVDA",
      "TSLA",
      "AMD",
      "COIN",
      "PLTR",
      "MSTR",
      "SMCI",
      "IONQ",
      "RKLB",
      "SHOP",
    ];
    const results = await Promise.allSettled(
      symbols.map((sym) =>
        get(`https://finnhub.io/api/v1/quote?symbol=${sym}&token=${KEY}`).then(
          (q) => ({
            ticker: sym,
            name: sym,
            price: safe(q.c),
            change: safe(q.c && q.pc ? ((q.c - q.pc) / q.pc) * 100 : null),
          }),
        ),
      ),
    );
    const all = results
      .filter((r) => r.status === "fulfilled" && r.value.price)
      .map((r) => r.value);
    const gainers = all
      .filter((r) => (r.change || 0) > 0)
      .sort((a, b) => b.change - a.change)
      .slice(0, 5);
    const losers = all
      .filter((r) => (r.change || 0) < 0)
      .sort((a, b) => a.change - b.change)
      .slice(0, 5);
    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({ gainers, losers }),
    };
  } catch (e) {
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ gainers: [], losers: [], error: e.message }),
    };
  }
};
